lib/ngtsc/workspace.symbols.ts

```diff
@@ -0,0 +1,516 @@
+  localMetaReader: LocalMetadataRegistry;
+  refEmitter: ReferenceEmitter;
+  referencesRegistry: ReferencesRegistry;
+  routeAnalyzer: NgModuleRouteAnalyzer;
+
+  resourceManager: AdapterResourceLoader;
+  moduleResolver: ModuleResolver;
+  cycleAnalyzer: CycleAnalyzer;
+  incrementalDriver: IncrementalDriver;
+  dtsTransforms: DtsTransformRegistry;
+  mwpScanner: ModuleWithProvidersScanner;
+}
+
+// code from :
+// https://github.com/angular/angular/blob/9.1.x/packages/compiler-cli/src/ngtsc/core/src/compiler.ts#L821
+class ReferenceGraphAdapter implements ReferencesRegistry {
+  constructor(private graph: ReferenceGraph) {}
+
+  add(source: Declaration, ...references: Reference<Declaration>[]): void {
+    for (const {node} of references) {
+      let sourceFile = node.getSourceFile();
+      if (sourceFile === undefined) {
+        sourceFile = getOriginalNode(node).getSourceFile();
+      }
+
+      // Only record local references (not references into .d.ts files).
+      if (sourceFile === undefined || !isDtsPath(sourceFile.fileName)) {
+        this.graph.add(source, node);
+      }
+    }
+  }
+}
+
+// All the code here comes from the ngtsc Compiler file, for more detail see :
+// https://github.com/angular/angular/blob/9.1.x/packages/compiler-cli/src/ngtsc/core/src/compiler.ts
+
+export class WorkspaceSymbols {
+  private options: NgCompilerOptions;
+  private rootNames: string[];
+  private toolkit: Partial<Toolkit> = {};
+  private isCore = false;
+  private analysed = false;
+  private oldProgram: Program
+
+  constructor(
+    private tsconfigPath: string,
+    private fs: FileSystem = new NodeJSFileSystem(),
+    private perfRecorder: PerfRecorder = NOOP_PERF_RECORDER
+  ) {
+    const config = readConfiguration(this.tsconfigPath);
+    this.options = config.options;
+    this.rootNames = config.rootNames;
+  }
+
+  // ------ PUBLIC API ----- //
+
+  /** Angular wrapper around the typescript host compiler */
+  // TODO: add reusable program
+  get host() {
+    return this.lazy('host', () => {
+      const baseHost = new NgtscCompilerHost(this.fs, this.options);
+      return NgCompilerHost.wrap(baseHost, this.rootNames, this.options, this.oldProgram || null);
+    });
+  }
+
+
+  /** Typescript program */
+  get program() {
+    return this.lazy('program', () => createProgram({
+        host: this.host,
+        rootNames: this.host.inputFiles,
+        options: this.options
+      })
+    );
+  }
+
+  /** Process all classes in the program */
+  get traitCompiler() {
+    return this.lazy('traitCompiler', () => new NgastTraitCompiler(
+        [this.cmptHandler, this.directiveHandler as any, this.pipeHandler, this.injectableHandler, this.moduleHandler],
+        this.reflector,
+        this.perfRecorder,
+        this.incrementalDriver,
+        this.options.compileNonExportedClasses !== false,
+        this.dtsTransforms
+      )
+    );
+  }
+
+  /** Handler for @Injectable() annotations */
+  get injectableHandler() {
```

Comment:
I'd suggest making these private. We don't need to expose a large public interface.